crates/wasmtime/src/runtime/vm/libcalls.rs

```diff
@@ -1158,6 +1192,30 @@ unsafe fn check_free(_store: &mut dyn VMStore, instance: &mut Instance, addr: u3
     Ok(0)
 }
 
+#[cfg(feature = "wmemcheck")]
+fn log_allocation_previous_to(instance: &mut Instance, addr: usize) {
+    if let Some(wmemcheck_state) = &mut instance.wmemcheck_state {
+        if let Some((prev_malloc, prev_len)) = wmemcheck_state.malloc_previous_to(addr) {
+            println!("previous malloc'd range was {:#x}..{:#x}", prev_malloc, prev_malloc + prev_len);
+            for (_, entry) in instance.exports() {
+                if let wasmtime_environ::EntityIndex::Memory(mem_index) = entry {
+                    let mem = instance.get_memory(*mem_index);
+                    for i in 0..prev_len {
+                        if i > 0 && i % 40 == 0 {
+                            println!();
+                        }
+                        unsafe {
+                            print!("{:02x} ", *mem.base.offset((prev_malloc + i) as isize));
+                        }
```

Comment:
This looks like code that could be useful for debugging, but as above with the `posix_memalign` out-pointer fetch, the memory accesses are unsafe, and also dumping the contents of previously allocated blocks might be too verbose. I think I'd prefer to omit this code for now; ultimately the right answer for any problem that needs examination of memory state like this is probably using a debugger (we have better support planned!) possibly with breakpoints on allocation events or trap-to-debugger on wmemcheck errors.